### PR TITLE
Set default load value for decimal values

### LINKED ISSUE #29113: The filler value '+0.0d' should be loaded as default value of decimal
**Description:**
$Subject

**Related Issues (optional):**
#28959

**Suggested Labels (optional):**
Priority/High, Type/Bug, Team/jBallerina

**Suggested Assignees (optional):**
gabilang

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #29113

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java [hunk 1]
```diff
@@ -188,6 +188,11 @@ public void testDefaultDecimalArgs() {
         BRunUtil.invoke(result, "testDefaultDecimalArgs");
     }
 
+    @Test(description = "Test decimal default values with InterOp functions")
+    public void testDefaultDecimalArgs2() {
+        BRunUtil.invoke(result, "testDefaultDecimalArgs2");
+    }
+
```
Comment (riyafa) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java:195]: Shall we move these function to the data provider?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java [hunk 2]
```diff
@@ -649,4 +649,13 @@ public static BDecimal defaultDecimalArgs(BDecimal a, BDecimal b) {
         BDecimal c = a.add(b);
         return c;
     }
+
+    public static Object defaultDecimalArgs2(String s, BDecimal d) {
```
Comment (riyafa) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java:653]: Shall we not name functions as 1 and 2? We can say the previous function as `defaultDecimalArgsAddition` and the second one as `defaultDecimalArgsMultiplication`? I think it is best to give some meaning to the name

File: tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/static_method_tests.bal [hunk 3]
```diff
@@ -155,6 +155,13 @@ public function testDefaultDecimalArgs() {
     test:assertEquals(val, expected);
 }
 
+public function testDefaultDecimalArgs2() {
```
Comment (riyafa) [inline tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/static_method_tests.bal:158]: shall we rename as discussed above?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java [hunk 4]
```diff
@@ -238,6 +240,15 @@ private void loadDefaultValue(MethodVisitor mv, BType bType) {
...
                 mv.visitInsn(ACONST_NULL);
                 break;
+            case TypeTags.DECIMAL:
+                mv.visitTypeInsn(NEW, DECIMAL_VALUE);
+                mv.visitInsn(DUP);
+                mv.visitInsn(DCONST_0);
+                mv.visitMethodInsn(INVOKESTATIC, BIG_DECIMAL, VALUE_OF_METHOD, String.format("(D)L%s;", BIG_DECIMAL),
+                        false);
+                mv.visitMethodInsn(INVOKESPECIAL, DECIMAL_VALUE, JVM_INIT_METHOD, String.format("(L%s;)V", BIG_DECIMAL),
+                        false);
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:250]: why do we need a `NEW` instruction. isn't following enough
```suggestion
                mv.visitInsn(DCONST_0);
                mv.visitMethodInsn(INVOKESTATIC, BIG_DECIMAL, VALUE_OF_METHOD, String.format("(D)L%s;", BIG_DECIMAL),
                        false);
```

**Review by manuranga — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java (hunk 4 above)
Comment (gabilang) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java:250]: I've used ```new DecimalValue(BigDecimal.valueOf(0.0));``` code segment to generate the decimal value. So I'm needed to add `DecimalValue` path. Without those lines, program crashed. So we need those lines.

**Review by manuranga — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java [hunk 5]
```diff
@@ -645,8 +645,17 @@ public static Object getValue() {
...
 
-    public static BDecimal defaultDecimalArgs(BDecimal a, BDecimal b) {
+    public static BDecimal defaultDecimalArgsAddition(BDecimal a, BDecimal b) {
         BDecimal c = a.add(b);
         return c;
     }
+
+    public static Object defaultDecimalArgs(String s, BDecimal d) {
+        if (!d.booleanValue()) {
+            BDecimal a = new DecimalValue(s);
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java:655]: Shall we use api to create decimal values. ValueCreator for tests too

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java [hunk 6]
```diff
@@ -645,8 +645,17 @@ public static Object getValue() {
...
     }
 
-    public static BDecimal defaultDecimalArgs(BDecimal a, BDecimal b) {
+    public static BDecimal defaultDecimalArgsAddition(BDecimal a, BDecimal b) {
         BDecimal c = a.add(b);
         return c;
     }
+
+    public static Object defaultDecimalArgs(String s, BDecimal d) {
+        if (!d.booleanValue()) {
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java:654]: Why need to check boolean here?

File: tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/static_method_tests.bal [hunk 7]
```diff
@@ -149,12 +149,19 @@ public function testGetDefaultValueWithBEnv() {
...
 
-public function testDefaultDecimalArgs() {
-    decimal val = defaultDecimalArgs(5);
+public function testDefaultDecimalArgsAddition() {
+    decimal val = defaultDecimalArgsAddition(5);
     decimal expected = 15.05;
     test:assertEquals(val, expected);
 }
 
+public function testDefaultDecimalArgs() {
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/javainterop/basic/static_method_tests.bal:158]: Shall we add a test for all simple types with defaultable paramter in functions_with_default_parameters.bal

ex. funtion foo(string a = "eee", int b = 2, ....) {

int c = b + 6;// add some operation
}

**Review by warunalakshitha — CHANGES_REQUESTED** [review]

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java (hunk 6 above)
Comment (gabilang) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java:654]: As per a random test, I wanted to ensure that the `BDecimal` is passed correctly. Is it seems more irrelevant?

Comment (warunalakshitha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/nativeimpl/jvm/tests/StaticMethods.java:654]: But it is always true based on your test?

**Review by warunalakshitha — APPROVED** [review]

**Comment by warunalakshitha** [comment]
I am merging this PR since Please send a new PR for test suggestion.

**Pull request merged**